File: scripts/example_basedata_fields.jl
#!/usr/bin/env julia
"""
Example: BaseData and Field Usage in Julia

This script demonstrates:
1. Creating BaseData objects with various tensor ranks
2. Saving BaseData to HDF5
3. Loading BaseData from HDF5
4. Creating Field objects from data
5. Saving Field objects
6. Loading Field objects
7. Evaluating fields at arbitrary points
"""

# Add the Firefly package
push!(LOAD_PATH, joinpath(@__DIR__, "..", "jlpkg"))

using Firefly

function example_scalar_field()
    """Example 1: Scalar complex field with frequency dependence"""
    println("\n" * "="^60)
    println("Example 1: Scalar Complex Field (inds=[])")
    println("="^60)

    # Parameters
    nk = 20  # Number of k-points
    nw = 10  # Number of frequency points

    # Create scalar data (w-k ordering: frequency varies slowest)
    # Each point has a complex value
    data = ComplexF32[]
    for iw in 1:nw
        for ik in 1:nk
            # Value depends on position
            real_part = Float32((ik-1) / nk)
            imag_part = Float32((iw-1) / nw)
            push!(data, complex(real_part, imag_part))
        end
    end

    # Save using Firefly's save_data_scalar
    println("Creating scalar field...")
    mesh = Int32[nk]
    domain = Float32[1.0]  # 1D BZ spanning [-0.5, 0.5]
    w_points = Float32[(i-1) / nw for i in 1:nw]

    filename = "/tmp/example_scalar_field_jl.h5"
    save_data_scalar(filename, data, mesh, domain, w_points, is_complex=true)
    println("Saved BaseData to: $filename")

    # Load BaseData
    println("Loading BaseData...")
    basedata = BaseData(filename)
    println("  is_complex: $(basedata.is_complex)")
    println("  inds: $(basedata.inds)")
    println("  mesh: $(basedata.mesh)")
    println("  nk: $(basedata.nk)")
    println("  nw: $(basedata.nw)")

    # Create Field from file
    println("Creating Field_C from file...")
    field = Field_C(filename)

    # Evaluate at a point
    println("Evaluating field at k=0.0, w=0.5...")
    k = Vec(0.0)  # Center of BZ
    w = 0.5
    value = field(k, w)
    println("  Result: $value")

    # Save Field (should be identical to BaseData save)
    field_filename = "/tmp/example_scalar_field_from_field_jl.h5"
    save(field, field_filename)
    println("Saved Field to: $field_filename")

    # Load Field again
    field2 = Field_C(field_filename)
    value2 = field2(k, w)
    println("Loaded Field evaluation: $value2")
    println("Values match: $(abs(value - value2) < 1e-6)")
end


function example_matrix_field()
    """Example 2: Matrix field (rank-2 tensor)"""
    println("\n" * "="^60)
    println("Example 2: Complex Matrix Field (inds=[3, 3])")
    println("="^60)

    # Parameters
    nk = 10  # Total k-points (10x10 mesh)
    mat_dim = 3  # 3x3 matrices

    # Create matrix data: vector of 3x3 matrices
    # Each matrix at each k-point
    data = Vector{Matrix{ComplexF32}}(undef, nk * nk)
    for ik in 1:(nk * nk)
        matrix = Matrix{ComplexF32}(undef, mat_dim, mat_dim)
        for i in 1:mat_dim
            for j in 1:mat_dim
                # Matrix elements depend on position and indices
                val = Float32((ik-1) / (nk*nk) + (i + j - 2) / 10.0)
                matrix[i, j] = complex(val, val / 10.0)
            end
        end
        data[ik] = matrix
    end

    # Save matrix field
    filename = "/tmp/example_matrix_field_jl.h5"
    mesh = Int32[nk, nk]
    domain = Float32[1.0 0.0; 0.0 1.0]  # 2D BZ
    w_points = Float32[]
    inds = Int32[mat_dim, mat_dim]

    println("Creating $(mat_dim)x$(mat_dim) matrix field...")
    save_data_matrix(filename, data, mesh, domain, w_points, inds, is_complex=true)
    println("Saved to: $filename")

    # Load and inspect BaseData
    println("Loading BaseData...")
    basedata = BaseData(filename)
    println("  inds: $(basedata.inds)")
    println("  rank: $(length(basedata.inds))")
    println("  dimension: $(basedata.dimension)")

    # Create Field_CM
    println("Creating Field_CM...")
    field = Field_CM(filename)

    # Evaluate at a point
    k = Vec(0.1, 0.2)
    println("Evaluating at k=($(k.x), $(k.y))...")
    matrix = field(k)
    println("  Result shape: $(size(matrix))")
    println("  Matrix[1,1] = $(matrix[1,1])")
    println("  Matrix[2,3] = $(matrix[2,3])")

    # Save and reload Field
    field_file = "/tmp/example_matrix_field_from_field_jl.h5"
    save(field, field_file)
    field2 = Field_CM(field_file)
    matrix2 = field2(k)
    println("Round-trip match: $(abs(matrix[1,1] - matrix2[1,1]) < 1e-6)")
end


function example_nonuniform_matrix()
    """Example 3: Non-uniform matrix (2x3)"""
    println("\n" * "="^60)
    println("Example 3: Non-Uniform Matrix Field (inds=[2, 3])")
    println("="^60)

    # Create 2x3 matrix data
    nk = 15
    dim1, dim2 = 2, 3

    data = Vector{Matrix{ComplexF32}}(undef, nk)
    for ik in 1:nk
        matrix = Matrix{ComplexF32}(undef, dim1, dim2)
        for i in 1:dim1
            for j in 1:dim2
                val = Float32(((ik-1) * dim1 * dim2 + (i-1) * dim2 + (j-1)) / (nk * dim1 * dim2))
                matrix[i, j] = complex(val, -val)
            end
        end
        data[ik] = matrix
    end

    filename = "/tmp/example_nonuniform_matrix_jl.h5"
    mesh = Int32[nk]
    domain = Float32[1.0]
    w_points = Float32[]
    inds = Int32[dim1, dim2]

    println("Creating $(dim1)x$(dim2) non-uniform matrix field...")
    save_data_matrix(filename, data, mesh, domain, w_points, inds, is_complex=true)

    # Load and use
    field = Field_CM(filename)
    k = Vec(0.0)
    result = field(k)
    println("  Result shape: $(size(result))")
    println("  Non-square matrix verified!")
end


function example_4d_vertex()
    """Example 4: 4D vertex tensor (multi-orbital interaction)"""
    println("\n" * "="^60)
    println("Example 4: 4D Vertex Tensor (inds=[2, 2, 2, 2])")
    println("="^60)

    # Create 2x2x2x2 tensor data (2-orbital vertex)
    nk = 25  # 5x5 k-mesh
    dim = 2
    nw = 3

    # Create 4D tensor data
    data = Vector{Array{ComplexF32, 4}}(undef, nw * nk)
    idx_global = 1
    for iw in 1:nw
        for ik in 1:nk
            tensor = Array{ComplexF32, 4}(undef, dim, dim, dim, dim)
            for i in 1:dim
                for j in 1:dim
                    for k in 1:dim
                        for l in 1:dim
                            idx = (i-1)*dim^3 + (j-1)*dim^2 + (k-1)*dim + (l-1)
                            val = Float32((ik + iw - 2) / (nk + nw) + idx / 100.0)
                            tensor[i, j, k, l] = complex(val, val/20.0)
                        end
                    end
                end
            end
            data[idx_global] = tensor
            idx_global += 1
        end
    end

    filename = "/tmp/example_4d_vertex_jl.h5"
    mesh = Int32[5, 5]
    domain = Float32[6.28 0.0; 0.0 6.28]  # Full BZ
    w_points = Float32[-0.1, 0.0, 0.1]
    inds = Int32[dim, dim, dim, dim]

    println("Creating $(dim)x$(dim)x$(dim)x$(dim) vertex tensor...")
    save_data_tensor4(filename, data, mesh, domain, w_points, inds, is_complex=true)

    # Load as Field_CM
    field = Field_CM(filename)
    k = Vec(3.14, 3.14)  # BZ center
    w = 0.0

    # Field_CM flattens 4D tensor to [d1*d2][d3*d4] matrix
    result = field(k, w)
    println("  Result shape (flattened): $(size(result))")
    println("  Expected: [$(dim*dim)][$(dim*dim)] = [4][4]")
    println("  result[1,1] = $(result[1,1])")
end


function example_vector_field()
    """Example 5: Vector field (rank-1)"""
    println("\n" * "="^60)
    println("Example 5: Vector Field (inds=[4])")
    println("="^60)

    # Create 4-component vector field
    nk = 10
    vec_dim = 4

    data = Vector{Vector{ComplexF32}}(undef, nk)
    for ik in 1:nk
        vector = Vector{ComplexF32}(undef, vec_dim)
        for i in 1:vec_dim
            val = Float32((ik-1) / nk + (i-1) / vec_dim)
            vector[i] = complex(val, -val/2.0)
        end
        data[ik] = vector
    end

    filename = "/tmp/example_vector_field_jl.h5"
    mesh = Int32[nk]
    domain = Float32[1.0]
    w_points = Float32[]
    inds = Int32[vec_dim]

    println("Creating $(vec_dim)-component vector field...")
    save_data_vector(filename, data, mesh, domain, w_points, inds, is_complex=true)

    # Load as Field_CM
    # For rank-1, Field_CM wraps as column matrix [vec_dim][1]
    field = Field_CM(filename)
    k = Vec(0.5)
    result = field(k)

    println("  Result shape (as column matrix): $(size(result))")
    println("  Vector components:")
    for i in 1:size(result, 1)
        println("    [$i]: $(result[i,1])")
    end
end


function main()
    """Run all examples"""
    println("\n" * "="^60)
    println("BaseData and Field Examples - Julia")
    println("="^60)

    example_scalar_field()
    example_matrix_field()
    example_nonuniform_matrix()
    example_4d_vertex()
    example_vector_field()

    println("\n" * "="^60)
    println("All examples completed successfully!")
    println("="^60)
    println("\nFiles created in /tmp/:")
    println("  - example_scalar_field_jl.h5")
    println("  - example_matrix_field_jl.h5")
    println("  - example_nonuniform_matrix_jl.h5")
    println("  - example_4d_vertex_jl.h5")
    println("  - example_vector_field_jl.h5")
    println("\nYou can inspect these files with h5dump or h5ls")
end


# Run if executed as script
if abspath(PROGRAM_FILE) == @__FILE__
    main()
end

File: scripts/example_basedata_fields.py
#!/usr/bin/env python3
"""
Example: BaseData and Field Usage in Python

This script demonstrates:
1. Creating BaseData objects with various tensor ranks
2. Saving BaseData to HDF5
3. Loading BaseData from HDF5
4. Creating Field objects from data
5. Saving Field objects
6. Loading Field objects
7. Evaluating fields at arbitrary points
"""

import sys
import os
import numpy as np

# Add the firefly package to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'pypkg'))

try:
    import firefly.src.module.imports.cpp_imports as ff
except ImportError as e:
    print(f"Error importing firefly: {e}")
    print("Make sure you've built the project with ./scripts/fly-build.sh")
    sys.exit(1)

def example_scalar_field():
    """Example 1: Scalar complex field with frequency dependence"""
    print("\n" + "="*60)
    print("Example 1: Scalar Complex Field (inds=[])")
    print("="*60)

    # Parameters
    nk = 20  # Number of k-points
    nw = 10  # Number of frequency points

    # Create scalar data (w-k ordering: frequency varies slowest)
    # Each point has a complex value
    data = np.zeros(nk * nw, dtype=np.complex64)
    idx = 0
    for iw in range(nw):
        for ik in range(nk):
            # Value depends on position
            real_part = float(ik) / nk
            imag_part = float(iw) / nw
            data[idx] = complex(real_part, imag_part)
            idx += 1

    # Save using firefly's save_data_scalar
    print("Creating scalar field...")
    mesh = np.array([nk], dtype=np.int32)
    domain = np.array([[1.0]], dtype=np.float32)  # 1D BZ spanning [-0.5, 0.5]
    w_points = np.array([float(i) / nw for i in range(nw)], dtype=np.float32)

    filename = "/tmp/example_scalar_field.h5"
    ff.save_data_scalar(filename, data, True, mesh, domain, w_points)
    print(f"Saved BaseData to: {filename}")

    # Load BaseData
    print("Loading BaseData...")
    basedata = ff.BaseData(filename)
    print(f"  is_complex: {basedata.is_complex}")
    print(f"  inds: {basedata.inds}")
    print(f"  mesh: {basedata.mesh}")
    print(f"  nk: {basedata.nk}")
    print(f"  nw: {basedata.nw}")

    # Create Field from file
    print("Creating Field_C from file...")
    field = ff.Field_C(filename)

    # Evaluate at a point
    print("Evaluating field at k=0.0, w=0.5...")
    k = ff.Vec(0.0)  # Center of BZ
    w = 0.5
    value = field(k, w)
    print(f"  Result: {value}")

    # Save Field (should be identical to BaseData save)
    field_filename = "/tmp/example_scalar_field_from_field.h5"
    field.save(field_filename)
    print(f"Saved Field to: {field_filename}")

    # Load Field again
    field2 = ff.Field_C(field_filename)
    value2 = field2(k, w)
    print(f"Loaded Field evaluation: {value2}")
    print(f"Values match: {abs(value - value2) < 1e-6}")


def example_matrix_field():
    """Example 2: Matrix field (rank-2 tensor)"""
    print("\n" + "="*60)
    print("Example 2: Complex Matrix Field (inds={3, 3})")
    print("="*60)

    # Parameters
    nk = 10  # Total k-points (10x10 mesh)
    mat_dim = 3  # 3x3 matrices

    # Create matrix data: vector of 3x3 matrices
    # Each matrix at each k-point
    data = []
    for ik in range(nk * nk):
        matrix = []
        for i in range(mat_dim):
            row = []
            for j in range(mat_dim):
                # Matrix elements depend on position and indices
                val = float(ik) / (nk*nk) + float(i + j) / 10.0
                row.append(complex(val, val / 10.0))
            matrix.append(row)
        data.append(matrix)

    # Save matrix field
    filename = "/tmp/example_matrix_field.h5"
    mesh = [nk, nk]
    domain = [[1.0, 0.0], [0.0, 1.0]]  # 2D BZ
    inds = [mat_dim, mat_dim]

    print(f"Creating {mat_dim}x{mat_dim} matrix field...")
    ff.save_data_matrix(filename, data, True, mesh, domain, [], inds)
    print(f"Saved to: {filename}")

    # Load and inspect BaseData
    print("Loading BaseData...")
    basedata = ff.BaseData(filename)
    print(f"  inds: {basedata.inds}")
    print(f"  rank: {len(basedata.inds)}")
    print(f"  dimension: {basedata.dimension}")

    # Create Field_CM
    print("Creating Field_CM...")
    field = ff.Field_CM(filename)

    # Evaluate at a point
    k = ff.Vec(0.1, 0.2)
    print(f"Evaluating at k=({k.x}, {k.y})...")
    matrix = field(k)
    print(f"  Result shape: {len(matrix)}x{len(matrix[0])}")
    print(f"  Matrix[0][0] = {matrix[0][0]}")
    print(f"  Matrix[1][2] = {matrix[1][2]}")

    # Save and reload Field
    field_file = "/tmp/example_matrix_field_from_field.h5"
    field.save(field_file)
    field2 = ff.Field_CM(field_file)
    matrix2 = field2(k)
    print(f"Round-trip match: {abs(matrix[0][0] - matrix2[0][0]) < 1e-6}")


def example_nonuniform_matrix():
    """Example 3: Non-uniform matrix (2x3)"""
    print("\n" + "="*60)
    print("Example 3: Non-Uniform Matrix Field (inds={2, 3})")
    print("="*60)

    # Create 2x3 matrix data
    nk = 15
    dim1, dim2 = 2, 3

    data = []
    for ik in range(nk):
        matrix = []
        for i in range(dim1):
            row = []
            for j in range(dim2):
                val = float(ik * dim1 * dim2 + i * dim2 + j) / (nk * dim1 * dim2)
                row.append(complex(val, -val))
            matrix.append(row)
        data.append(matrix)

    filename = "/tmp/example_nonuniform_matrix.h5"
    mesh = [nk]
    domain = [[1.0]]
    inds = [dim1, dim2]

    print(f"Creating {dim1}x{dim2} non-uniform matrix field...")
    ff.save_data_matrix(filename, data, True, mesh, domain, [], inds)

    # Load and use
    field = ff.Field_CM(filename)
    k = ff.Vec(0.0)
    result = field(k)
    print(f"  Result shape: {len(result)}x{len(result[0])}")
    print(f"  Non-square matrix verified!")


def example_4d_vertex():
    """Example 4: 4D vertex tensor (multi-orbital interaction)"""
    print("\n" + "="*60)
    print("Example 4: 4D Vertex Tensor (inds={2, 2, 2, 2})")
    print("="*60)

    # Create 2x2x2x2 tensor data (2-orbital vertex)
    nk = 25  # 5x5 k-mesh
    dim = 2
    nw = 3

    # Create 4D tensor data
    # Total size: nw * nk * (2*2*2*2) = 3 * 25 * 16 = 1200 values
    data = []
    for iw in range(nw):
        for ik in range(nk):
            tensor = []
            for i in range(dim):
                t1 = []
                for j in range(dim):
                    t2 = []
                    for k in range(dim):
                        t3 = []
                        for l in range(dim):
                            idx = i*dim**3 + j*dim**2 + k*dim + l
                            val = float(ik + iw) / (nk + nw) + float(idx) / 100.0
                            t3.append(complex(val, val/20.0))
                        t2.append(t3)
                    t1.append(t2)
                tensor.append(t1)
            data.append(tensor)

    filename = "/tmp/example_4d_vertex.h5"
    mesh = [5, 5]
    domain = [[6.28, 0.0], [0.0, 6.28]]  # Full BZ
    w_points = [-0.1, 0.0, 0.1]
    inds = [dim, dim, dim, dim]

    print(f"Creating {dim}x{dim}x{dim}x{dim} vertex tensor...")
    ff.save_data_tensor4(filename, data, True, mesh, domain, w_points, inds)

    # Load as Field_CM
    field = ff.Field_CM(filename)
    k = ff.Vec(3.14, 3.14)  # BZ center
    w = 0.0

    # Field_CM flattens 4D tensor to [d1*d2][d3*d4] matrix
    result = field(k, w)
    print(f"  Result shape (flattened): {len(result)}x{len(result[0])}")
    print(f"  Expected: [{dim*dim}][{dim*dim}] = [4][4]")
    print(f"  result[0][0] = {result[0][0]}")


def example_vector_field():
    """Example 5: Vector field (rank-1)"""
    print("\n" + "="*60)
    print("Example 5: Vector Field (inds={4})")
    print("="*60)

    # Create 4-component vector field
    nk = 10
    vec_dim = 4

    data = []
    for ik in range(nk):
        vector = []
        for i in range(vec_dim):
            val = float(ik) / nk + float(i) / vec_dim
            vector.append(complex(val, -val/2.0))
        data.append(vector)

    filename = "/tmp/example_vector_field.h5"
    mesh = [nk]
    domain = [[1.0]]
    inds = [vec_dim]

    print(f"Creating {vec_dim}-component vector field...")
    ff.save_data_vector(filename, data, True, mesh, domain, [], inds)

    # Load as Field_CM
    # For rank-1, Field_CM wraps as column matrix [vec_dim][1]
    field = ff.Field_CM(filename)
    k = ff.Vec(0.5)
    result = field(k)

    print(f"  Result shape (as column matrix): {len(result)}x{len(result[0])}")
    print(f"  Vector components:")
    for i in range(len(result)):
        print(f"    [{i}]: {result[i][0]}")


def main():
    """Run all examples"""
    print("\n" + "="*60)
    print("BaseData and Field Examples - Python")
    print("="*60)

    example_scalar_field()
    example_matrix_field()
    example_nonuniform_matrix()
    example_4d_vertex()
    example_vector_field()

    print("\n" + "="*60)
    print("All examples completed successfully!")
    print("="*60)
    print("\nFiles created in /tmp/:")
    print("  - example_scalar_field.h5")
    print("  - example_matrix_field.h5")
    print("  - example_nonuniform_matrix.h5")
    print("  - example_4d_vertex.h5")
    print("  - example_vector_field.h5")
    print("\nYou can inspect these files with h5dump or h5ls")


if __name__ == "__main__":
    main()
